web/pages/[username]/[contractSlug].tsx

```diff
@@ -203,31 +209,25 @@ export function ContractPageContent(props: FirstArgument<typeof ContractPage>) {
   )
 }
 
-function ContractLeaderboard(props: { contract: Contract; bets: Bet[] }) {
-  const { contract, bets } = props
+function ContractLeaderboard(props: { userProfits: { [id: string]: number } }) {
+  const { userProfits } = props
   const [users, setUsers] = useState<User[]>()
 
-  // Create a map of userIds to total profits (including sales)
-  const betsByUser = _.groupBy(bets, 'userId')
-  const userProfits = _.mapValues(betsByUser, (bets) =>
-    _.sumBy(bets, (bet) => resolvedPayout(contract, bet) - bet.amount)
-  )
-
-  // Find the 5 users with the most profits
-  const top5Ids = _.entries(userProfits)
-    .sort(([i1, p1], [i2, p2]) => p2 - p1)
-    .filter(([, p]) => p > 0)
-    .slice(0, 5)
-    .map(([id]) => id)
-
   useEffect(() => {
+    // Find the 5 users with the most profits
+    const top5Ids = _.entries(userProfits)
+      .sort(([i1, p1], [i2, p2]) => p2 - p1)
+      .filter(([, p]) => p > 0)
+      .slice(0, 5)
+      .map(([id]) => id)
+
     if (top5Ids.length > 0) {
       listUsers(top5Ids).then((users) => {
         const sortedUsers = _.sortBy(users, (user) => -userProfits[user.id])
         setUsers(sortedUsers)
       })
     }
-  }, [])
+  }, [userProfits])
```

Comment:
See also my above comment if you wanted to optimize away this re-fetching.

Reply:
Hmm...you're right. I'm still not super happy. And I'm not happy with the `useMemo` fix, because it's spooky action at a distance (the reason we have a `useMemo` up here, is that down somewhere in the component tree, where you're not looking right now, something has a `useEffect` taking this as a dependency...)

But it works OK (the parent component is not getting re-rendered much) so in the interest of time I will merge this for now and maybe think about it later.

Thanks a lot for your feedback, this is extremely helpful for improving my ability to produce high-quality React programs involving hooks.